codec/src/main/java/io/netty/handler/codec/compression/Lz4FrameEncoder.java

```diff
@@ -163,38 +177,77 @@ private static int compressionLevel(int blockSize) {
     }
 
     @Override
+    protected ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect) {
+        return  allocateBuffer(ctx, msg, preferDirect, true);
+    }
+
+    ByteBuf allocateBuffer(ChannelHandlerContext ctx, ByteBuf msg, boolean preferDirect,
+                                     boolean allowEmptyReturn) {
+        int targetBufSize = 0;
+        int remaining = msg.readableBytes() + buffer.readableBytes();
+        while (remaining > 0) {
+            int curSize = Math.min(blockSize, remaining);
+            remaining -= curSize;
+            // calculate the total compressed size of the current block (including header) and add to the total
+            targetBufSize += compressor.maxCompressedLength(curSize) + HEADER_LENGTH;
+        }
+
+        if (targetBufSize > maxEncodeSize || 0 > targetBufSize) {
+            String errorMsg = "requested encode buffer size (%d bytes) exceeds the maximum allowable size (%d bytes)";
+            throw new EncoderException(String.format(errorMsg, targetBufSize, maxEncodeSize));
+        }
+
+        if (allowEmptyReturn && targetBufSize < blockSize) {
+            //would prefer to return null or Unpooled.EMPTY_BUFFER, but MessageToByteEncoder calls buffer.release()
+            return ctx.alloc().heapBuffer(0, 0);
+        }
+
+        if (preferDirect) {
+            return ctx.alloc().ioBuffer(targetBufSize, targetBufSize);
+        } else {
+            return ctx.alloc().heapBuffer(targetBufSize, targetBufSize);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Encodes the input buffer into {@link #blockSize} chunks in the output buffer. Data is only compressed and
+     * written once we hit the {@link #blockSize}; else, it is copied into the backing {@link #buffer} to await
+     * more data.
+     */
+    @Override
     protected void encode(ChannelHandlerContext ctx, ByteBuf in, ByteBuf out) throws Exception {
         if (finished) {
             out.writeBytes(in);
             return;
         }
 
-        int length = in.readableBytes();
-
         final ByteBuf buffer = this.buffer;
-        final int blockSize = buffer.capacity();
-        while (currentBlockLength + length >= blockSize) {
-            final int tail = blockSize - currentBlockLength;
-            in.getBytes(in.readerIndex(), buffer, currentBlockLength, tail);
-            currentBlockLength = blockSize;
-            flushBufferedData(out);
-            in.skipBytes(tail);
-            length -= tail;
+        int length = in.readableBytes();
+        while (length > 0) {
+            final int nextChunkSize = Math.min(length, buffer.writableBytes());
+            in.readBytes(buffer, nextChunkSize);
+            length -= nextChunkSize;
+
+            if (!buffer.isWritable()) {
+                flushBufferedData(out);
+                buffer.clear();
```

Comment:
should we instead just have `flushBufferedData` be responsible for setting the indexes on `buffer` according to how much was consumed?